pkg/controller/secrets/tenant_controller.go

```diff
@@ -0,0 +1,184 @@
+	}
+	r.status.OnCRFound()
+
+	// Get all Tenants so we can perform validation.
+	tenants := operatorv1.TenantList{}
+	if err = r.client.List(ctx, &tenants); err != nil {
+		return reconcile.Result{}, err
+	}
+	for _, t := range tenants.Items {
+		if t.Spec.ID == tenant.Spec.ID && t.Namespace != tenant.Namespace {
+			// A tenant in a different namespace has the same ID as this tenant. This is not allowed.
+			r.status.SetDegraded(operatorv1.ResourceReadError, fmt.Sprintf("Multiple tenants with the ID '%s'", t.Spec.ID), err, logc)
+			return reconcile.Result{}, nil
+		}
+		if t.Namespace == tenant.Namespace && t.Name != tenant.Name {
+			// Multiple tenants in the same namespace. This is not allowed.
+			r.status.SetDegraded(operatorv1.ResourceReadError, fmt.Sprintf("Multiple Tenants in namespace '%s'", t.Namespace), err, logc)
+			return reconcile.Result{}, nil
+		}
+	}
+	// Get Installation resource.
+	_, installation, err := utils.GetInstallation(context.Background(), r.client)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.status.SetDegraded(operatorv1.ResourceNotFound, "Installation not found", err, logc)
+			return reconcile.Result{}, err
+		}
+		r.status.SetDegraded(operatorv1.ResourceReadError, "An error occurred while querying Installation", err, logc)
+		return reconcile.Result{}, err
+	}
+
+	// Create a certificate manager for this tenant. This certificate manager will load the CA for this tenant, creating it if needed,
+	// and can be used to sign any certificates needed for this tenant's components.
+	opts := []certificatemanager.Option{
+		certificatemanager.AllowCACreation(),
+		certificatemanager.WithLogger(logc),
+		certificatemanager.WithTenant(tenant),
+	}
+	cm, err := certificatemanager.Create(r.client, installation, r.clusterDomain, tenant.Namespace, opts...)
+	if err != nil {
+		r.status.SetDegraded(operatorv1.ResourceCreateError, "Unable to create CA", err, logc)
+		return reconcile.Result{}, err
+	}
+	cm.AddToStatusManager(r.status, tenant.Namespace)
+
+	// Create the CA in the tenant's namespace.
+	keyPairOptions := []rcertificatemanagement.KeyPairOption{
+		rcertificatemanagement.NewKeyPairOption(cm.KeyPair(), true, true),
+	}
+
+	// Get the cluster-scoped CA certificate.
+	clusterCA, err := cm.GetCertificate(r.client, certificatemanagement.CASecretName, common.OperatorNamespace())
+	if err != nil {
+		r.status.SetDegraded(operatorv1.ResourceReadError, "Error querying CA certificate", err, logc)
+		return reconcile.Result{}, err
+	}
+
+	// Create a trusted bundle for this tenant. This bundle is provided to the tenant pods so that they can verify
+	// each other's certificates. Each tenant needs to trust:
+	// - Certificates signed by its own CA
+	// - Certificates signed by the cluster-scoped Tigera CA
+	trustedBundle := cm.CreateTrustedBundle()
+	trustedBundle.AddCertificates(clusterCA)
+
+	// TODO: Provision a trusted bundle that includes system certificates for components that need it.
```

Comment:
This seems too important to address before merging.

Reply:
Nope - we're merging to prevent rotting of this branch vs. master. This codepath only impacts the multi-tenant scenario, which is largely working in this PR but still has several TODOs (including this one) before it is production ready.

The main goal is to get the refactor in and testing the mainline single-tenant path.